source/common/network/address_impl.h

```diff
@@ -165,6 +177,16 @@ class Ipv6Instance : public InstanceBase {
     return reinterpret_cast<const sockaddr*>(&ip_.ipv6_.address_);
   }
   socklen_t sockAddrLen() const override { return sizeof(sockaddr_in6); }
+  bool openSocket(std::string const& address) {
+    if (!address.empty()) {
+      if (1 != inet_pton(AF_INET6, address.c_str(), &ip_.ipv6_.address_.sin6_addr)) {
```

Comment:
It can, but it might be cleaner if you want to combine the factory methods per the above suggestion, to capture the context of the error message in the implementation.

After seeing that this class already is an implementation a pure interface, I think putting a pure virtual `open` method at this level looks like a really clean approach. The caller of the open() method (the factory function) does not know whether we are in ipv6 or ipv4 mode. So we'd want to return that as an absl::Status here.